ccf_transaction_processor/transaction_processor/pdo_tp.cpp

```diff
@@ -0,0 +1,624 @@
+/* Copyright 2020 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "pdo_tp.h"
+
+using namespace std;
+using namespace ccf;
+using namespace tls;
+
+
+namespace ccfapp
+{
+
+    TPHandler::TPHandler(Store& store):
+        UserHandlerRegistry(store),
+        enclavetable(store.create<string, EnclaveInfo>("encalves")),
+        contracttable(store.create<string, ContractInfo>("contracts")),
+        ccltable(store.create<string, ContractStateInfo>("ccl_updates")),
+        signer(store.create<string, map<string, string>>("signer"))
+    {
+        ledger_signer_local = NULL;
+    }
+
+    string TPHandler::sign_document(const string& document){
+        vector<uint8_t> doc_vector(document.begin(), document.end());
+        auto sign = ledger_signer_local->sign(doc_vector);
+        return b64_from_raw(sign.data(), sign.size());
+    }
+
+    void TPHandler::init_handlers(Store& store){
+
+        UserHandlerRegistry::init_handlers(store);
+
+        //======================================================================================================
+        // ping handler implementation
+        auto ping = [this](Store::Tx& tx, const nlohmann::json& params) {
+            return make_success(true);
+        };
+
+        //======================================================================================================
+        // get_ledger_key implementation
+        auto get_ledger_key = [this](Store::Tx& tx, const nlohmann::json& params) {
+            auto signer_view = tx.get_view(signer);
```

Comment:
view is a standard terminology used within CCF (https://microsoft.github.io/CCF/developers/kv/kv_how_to.html). So sticking with it. 

Regarding threading, please see https://microsoft.github.io/CCF/developers/threading.html.
Also, please see that we are using CCF release 0.7.1, which has initial threading support. I expect more changes once we move to recent releases. For now, it is best that we explicitly set "number of enclave threads to 1" during build. CCF permits this. I will make a note of this in the documentation. 

Finally, to answer your question, my understanding is that if two threads access the same view, one of them will fail if both try to update the view.

Reply:
> view is a standard terminology used within CCF. So sticking with it.

I wouldn't expect most folks (myself included) to be familiar with CCF terminology. My suggestion was simply to make the code a bit more generally understandable :)

> my understanding is that if two threads access the same view, one of them will fail if both try to update the view.

Interesting, so it seems like views internally handle locking. It might be helpful to point that out somewhere in the documentation.

Reply:
both ccf code & documentation are evolving rapidly. For now I am going to suggest using only one thread. Will clarify in documentation.